### PR TITLE
Types and RPC calls

### PR DESCRIPTION
**Pull request by philipstanislaus**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// most importantly RPC_URL for a custom endpoint.
+func NewDefaultConfig() Config {
+	c := Config{}
+	c.SetDefaultRPCURL()
+	return c
+}
+
+// SetDefaultRPCURL reads the env variable RPC_URL and sets it in the config. If that variable is unset or empty,
+// it will fallback to "http://127.0.0.1:9933"
+func (c *Config) SetDefaultRPCURL() {
```
Comment (vimukthi-git) [inline config/config.go]: Use of `Set` in a function name implies that it takes in a param. Wouldn't it make sense to name it `ExtractDefaultRPCURL`?

File: rpc/chain/chain_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+package chain
+
+import (
+	"os"
+	"testing"
+
+	"github.com/centrifuge/go-substrate-rpc-client/config"
```
Comment (vimukthi-git) [inline rpc/chain/chain_test.go]: org imports

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	\
+	export PATH="/usr/local/go/bin:$PATH"; \
+	go version
+
+ENV GOPATH /go
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
+ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
+WORKDIR $GOPATH
+
+# For debugging
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply Dockerfile]: Good point, removed.

File: main_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	gsrpc "github.com/centrifuge/go-substrate-rpc-client"
+)
+
+func Example_simpleConnect() {
+	api, err := gsrpc.NewSubstrateAPI(config.NewDefaultConfig().RPCURL)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(api.RPC.Chain.GetBlockHashLatest())
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply main_test.go]: I'm using Go's example feature here, it's a test in the end that compares fmt.Println with the provided comment in `// Output: 0xXXX`. Need to use your RPC mock though to have consistent outputs.

File: author.go.to_migrate [hunk 5]
```diff
@@ -213,24 +216,34 @@ func (e Extrinsic) Encode(encoder scale.Encoder) error {
 		// Immortal
 		Era: 0,
 	}
+
+	fmt.Println("sigPay 1", sigPay)
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply author.go.to_migrate]: Good point, done.

File: config/config.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	if url, ok := os.LookupEnv("RPC_URL"); ok {
+		c.RPCURL = url
+		return
+	}
+	// FIXME: due to a size limit, websocket connections don't work with getMetadata right now.
+	// 	Related issue: https://github.com/ethereum/go-ethereum/issues/16846
+	//  Should get fixed with https://github.com/ethereum/go-ethereum/pull/19866 , released in 1.9.1
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply config/config.go]: It has been merged, but our build depends on an older release. I'll try to upgrade shortly.

File: client/client.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+type client struct {
+	*rpc.Client
+
+	// metadataVersioned is the metadata cache to prevent unnecessary requests
+	//metadataVersioned *MetadataVersioned
+
+	//metadataLock sync.RWMutex
+}
+
+// TODO move to State struct
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply client/client.go]: Maybe it's subjective, but in my view, client belongs to the networking/access layer, while metadata cache belongs to a higher application layer. I like to keep these layers as separated as possible for easier swapping of modules and simplifying the architecture. What do you think about that?

File: config/config.go (hunk 1 above)
Comment (philipstanislaus) [reply config/config.go]: Totally, good catch.

Comment (philipstanislaus) [reply config/config.go]: Changed.

File: main_test.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+	//chain := api.RPC.System.Chain()
+	//name := api.RPC.System.Name()
+	//version := api.RPC.System.Version()
+
+	// TODO Output: 0x26735e602578d1f1741552e32d074a5fdefde5bf5e1ad074ba0ba6a96c9cc014
+}
+
+// listen to new blocks
+// func Example_listenToNewBlocks(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply main_test.go]: @vimukthi-git it's an example that will be added in the next milestone. I added a TODO to the comment.

File: main.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+package gsrpc
+
+import (
+	"github.com/centrifuge/go-substrate-rpc-client/client"
+	"github.com/centrifuge/go-substrate-rpc-client/rpc"
+)
+
+type SubstrateAPI struct {
+	RPC    *rpc.RPC
+	URL    string
```
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply main.go]: Good catch, moved it to the client struct

File: config/config.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipstanislaus) [reply config/config.go]: Upgraded.

File: main_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (vimukthi-git) [reply main_test.go]: Cool, I didn't know about this feature, Learnt something today :) 
But I think it makes sense add the RPC mock here.

File: client/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vimukthi-git) [reply client/client.go]: Yes it's for sure subjective. If you can make the `author` RPCs work with the metadata cache without passing the `state` RPC to `author` RPCs I'm fine with this change.

**Pull request merged**